File: strategy_pattern/go/duck/mallard_duck.go
package duck

import "duckapp/behavior"

type MallardDuck struct {
	Duck
}

func NewMallardDuck() *MallardDuck {
	mallard := &MallardDuck{}
	mallard.SetFlyBehavior(&behavior.FlyWithWings{})
	mallard.SetQuackBehavior(&behavior.Quack{})
	return mallard
}

func (m *MallardDuck) Display() {
	println("I'm a real Mallard duck")
}

File: strategy_pattern/ruby/ducks/mallard_duck.rb
require_relative '../behaviors/fly_behavior'
require_relative '../behaviors/quack_behavior'
require_relative 'duck'

class MallardDuck < Duck
  def initialize
    super
    @fly_behavior = FlyBehavior::FlyWithWings.new
    @quack_behavior = QuackBehavior::Quack.new
  end

  def display
    puts "I'm a real Mallard duck"
  end
end
